mlx_lm/generate.py

```diff
@@ -284,21 +285,56 @@ def maybe_quantize_kv_cache(prompt_cache, quantized_kv_start, kv_group_size, kv_
                 )
 
 
+class SearchStrategy(ABC):
+    """
+    A strategy for obtaining a sequence of tokens from the model.
+    """
+
+    @abstractmethod
+    def generate(
+        self,
+        prompt: mx.array,
+        prompt_cache: List[Any],
+        quantize_cache_fn: Callable[[Any], None],
+        total_prompt_tokens: int,
```

Comment:
I think it's a bit awkward that `generate` needs to take `total_prompt_tokens` and `prompt_progress_callback`, but I could immediately see a nice way around this.